### ISSUE TITLE
False positive no-magic-numbers enforceConst

### ISSUE THREAD

**Issue by vsemozhetbyt** [eslint/eslint#6388]
**What version of ESLint are you using?**

2.12.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**
```
"no-magic-numbers": [2, {
  "ignore": [-1, 0, 1, 2],
  "ignoreArrayIndexes": true,
  "enforceConst": true,
  "detectObjects": false,
}]
```
**What did you do? Please include the actual source code causing the issue.**
```js
'use strict';

let numberOfOperations = 5;

while (numberOfOperations--) {
  console.log('Hi');
}
```
**What did you expect to happen?**

eslint should check if numeric variable changes before applying `enforceConst`.

**What actually happened? Please include the actual, raw output from ESLint.**
```
  3:26  error  Number constants declarations must use 'const'  no-magic-numbers

✖ 1 problem (1 error, 0 warnings)
```

**Comment by michaelficarra** [comment]
You should write it like this instead:

```js
'use strict';

const totalNumberOfOperations = 5;

let numberOfOperations = totalNumberOfOperations;
while (numberOfOperations--) {
  console.log('Hi');
}
```

**Comment by vsemozhetbyt** [comment]
@michaelficarra What's the reason? The number gets its descriptive name so it is not magic anymore. Should I produce several variables just to make eslint happy? Does not this approach distort priority?

**Comment by alberto** [comment]
@vsemozhetbyt That's the purpose of `enforceConst`. If you don't want that behaviour, remove that option.

Maybe you want to use `prefer-const` instead.

**Comment by vsemozhetbyt** [comment]
`prefer-const` works as expected because it detect if the user intends the variable to change. If I intend some counter to change, why the `enforceConst` is triggered at all? It is widespread case for testing and benchmarking, why we should overcomplicate it? See, for example, in the V8 tests:

https://github.com/v8/v8/blob/master/test/mjsunit/regress/regress-113924.js

**Comment by alberto** [comment]
Because that's what enforceConst does: http://eslint.org/docs/rules/no-magic-numbers#enforceconst

If you want to be able to change number declarations you should not use `enforceConst` and use `var` or `let` instead. Additionally you can complement it with `prefer-const` if you want eslint to tell you if you are not really changing it and could use `const` instead.

**Comment by vsemozhetbyt** [comment]
Oh, I see now. I thought `enforceConst` is just particular type of `prefer-const`. Thank you. Sorry for the confusion.
